text/0000-dyn-assoc.md

```diff
@@ -0,0 +1,266 @@
+
+## Mentions of the `Self` type in methods and associated functions
+As [explained back in the guide section][mentions-of-the-implementing-type-in-assoc-fn-guide], `Self` as a concrete type is still not allowed in traits if these wish to stay object-safe. Trait objects of a trait in the declaration of that exact same trait are perfectly fine, and even [do compile in present Rust](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=45c35ab9274c22b323a9016796c635a5):
+```rust
+trait Thing {
+    fn make_a_friend(&self) -> Box<dyn Thing>;
+}
+struct SimpleThing;
+impl Thing for SimpleThing {
+    fn make_a_friend(&self) -> Box<dyn Thing> {
+        Box::from(SimpleThing)
+    }
+}
+struct BetterThing;
+impl Thing for BetterThing {
+    fn make_a_friend(&self) -> Box<dyn Thing> {
+        Box::from(BetterThing)
+    }
+}
+
+fn main() {
+    let sithing: Box<dyn Thing> = Box::from(SimpleThing);
+    let bething: Box<dyn Thing> = Box::from(BetterThing);
+    let sithing2 = sithing.make_a_friend();
+    let bething2 = sithing.make_a_friend();
+}
+```
+Supporting the actual `Self` type [might still be introduced in a later RFC][future-possibilities].
+
+## Edge cases
+None unresolved yet.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None known yet.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The constantly mentioned alternative is declaring the associated functions in traits which are meant to be trait objects with a `&self` parameter, effectively converting all associated functions into trait methods. This becomes a major source of confusion if the trait object is mutably borrowed elsewhere while this "`&self`ified" method is called. This introduces lots of unnecessary artificial curly-bracket scopes which hurt readability in most cases.
+
+That workaround is also wrong from a semantic standpoint, since trait methods by design should use the fields of the trait object in some way, either directly or indirectly. Again, an unused `&self` just to combat language limitations introduces confusion.
+
+Last but not least, "`&self`ifying" works terribly when the trait and the code creating and using trait objects of it are in different crates made by different people. In this case, the library developer has to add `&self` to the associated trait functions, hurting the code using the trait without trait object, which loses the ability to use its associated functions without an actual object of that trait. As a result, library developers would be forced to have the associated functions in their struct's main `impl` block and "`&self`ified" wrappers around these in trait implementations. This is incomprehensibly hacky and non-idiomatic, seriously hurting library design in the long run but inapparent in the beginning of development.
+
+# Prior art
+[prior-art]: #prior-art
```

Comment:
[GObject](https://en.wikipedia.org/wiki/GObject) (the runtime type system behind GTK, GStreamer, etc.) also allows for such things, which shouldn't be much of a surprise as the vtable is manually managed there.

You can put fields into the vtable (-> associated constants) or function pointers without an instance parameter (-> associated functions). This is used in quite a few places in GStreamer in practice.

[Vala](https://en.wikipedia.org/wiki/Vala_(programming_language)), a C#-inspired language on top of GObject that compiles down to C handles these as with the `class` keyword AFAIU. Not to be confused with `virtual`, which requires an instance parameter, or `static` which can't be overridden/implemented by subclasses / interface implementations but exists only exactly once.

```vala
public class MyClass {
    public class int some_field;
    public class void some_function() { ... }
}
```